### PR TITLE
[KT] Add support of uint64_t and int64_t types into esimd radix sort

### PR DESCRIPTION
**Pull request by SergeyKopienko**
In this PR we have added support of uint64_t and int64_t types into esimd radix sort:
- into one workgroup impl;
- into cooperative impl;
- still not implemented for onesweep impl.

Performance degradation not detected.

### CODE REVIEW

File: include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h [hunk 1]
```diff
@@ -436,6 +436,100 @@ VectorStore(uint32_t offset,
...
+    // +----------------------------------------------------------+-----------------------------------------------------------+-----+----------------------------------------------------------+
+    //                                                                                          |
+    // __uint32_mask, simd_mask<2 * N>:                                                         V
+    // +-----------------------------+----------------------------+-----------------------------+-----------------------------+-----+-----------------------------+----------------------------+
+    // |             m0              |             m0             |             m1              |             m1              | ... |             mN              |             mN             |
+    // +-----------------------------+----------------------------+-----------------------------+-----------------------------+-----+-----------------------------+----------------------------+
+
+    using TypeX32 = decltype(DestTypeResolver(T{}));
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h:502]: Can you use just `using TypeX32 = ::std::uint32_t`?

File: include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h [hunk 2]
```diff
@@ -436,6 +436,100 @@ VectorStore(uint32_t offset,
...
+    // simd select stride
+    #pragma unroll
+    for (size_t src_index = 0; src_index < N; ++src_index)
+    {
+        __x32_offsets[2 * src_index    ] = offsets[src_index];
+        __x32_offsets[2 * src_index + 1] = offsets[src_index] + sizeof(TypeX32);
+
+        __x32_mask[2 * src_index    ] = mask[src_index];
+        __x32_mask[2 * src_index + 1] = mask[src_index];
+    }
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h:521]: Have you tried using `select`?
```suggestion
    __x32_offsets.template select<N, 2>(0) = offsets[src_index];
    __x32_offsets.template select<N, 2>(1) = offsets[src_index] + sizeof(TypeX32);
    __x32_mask.template<select<N, 2>(0) = mask[src_index];
    __x32_mask.template<select<N, 2>(1) = mask[src_index];
```

File: include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h [hunk 3]
```diff
@@ -436,6 +436,100 @@ VectorStore(uint32_t offset,
...
+
+    // https://intel.github.io/llvm-docs/doxygen/group__sycl__esimd__memory__slm.html#gac6d34e476b1addadb3f27648d62322a4
+    //template<typename T , int N>
+    //__ESIMD_API std::enable_if_t<(N == 1 || N == 8 || N == 16 || N == 32) && (sizeof(T) <= 4)>
+    //sycl::_V1::ext::intel::esimd::slm_scatter(
+    //    simd<uint32_t, N> offsets,                    // Byte offsets within the SLM of each element.
+    //    simd<T, N> vals,                              // Values to be written.
+    //    simd_mask<N> mask = 1);                       // Operation mask. All-1 by default.
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h:529]: This comment looks excessive. Could you remove it?

File: test/parallel_api/experimental/radix_sort_esimd.pass.cpp [hunk 4]
```diff
@@ -370,6 +399,8 @@ int main()
             test_general_cases<uint16_t>(size);
             test_general_cases<int     >(size);
             test_general_cases<uint32_t>(size);
+            test_general_cases<uint64_t>(size);
+            test_general_cases<int64_t >(size);
```
Comment (dmitriy-sobolev) [inline test/parallel_api/experimental/radix_sort_esimd.pass.cpp:403]: Could you remove TODO notes about support of these types from `esimd_radix_sort.hpp`?

File: include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h [hunk 5]
```diff
@@ -436,6 +436,100 @@ VectorStore(uint32_t offset,
...
+            __ESIMD_NS::simd<T, N> vals,
+            __ESIMD_NS::simd_mask<N> mask = 1)
+{
+    // https://intel.github.io/llvm-docs/doxygen/group__sycl__esimd__memory__slm.html#gac6d34e476b1addadb3f27648d62322a4
+    //template<typename T , int N>
+    //__ESIMD_API std::enable_if_t<(N == 1 || N == 8 || N == 16 || N == 32) && (sizeof(T) <= 4)>
+    //sycl::_V1::ext::intel::esimd::slm_scatter(
+    //    simd<uint32_t, N> offsets,                    // Byte offsets within the SLM of each element.
+    //    simd<T, N> vals,                              // Values to be written.
+    //    simd_mask<N> mask = 1);                       // Operation mask. All-1 by default.
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h]: Fixed (removed)

File: include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h (hunk 1 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h:495]: Fixed.

File: include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h [hunk 6]
```diff
@@ -436,6 +436,100 @@ VectorStore(uint32_t offset,
...
+    //                                                                                          |
+    // __uint32_mask, simd_mask<2 * N>:                                                         V
+    // +-----------------------------+----------------------------+-----------------------------+-----------------------------+-----+-----------------------------+----------------------------+
+    // |             m0              |             m0             |             m1              |             m1              | ... |             mN              |             mN             |
+    // +-----------------------------+----------------------------+-----------------------------+-----------------------------+-----+-----------------------------+----------------------------+
+
+    using TypeX32 = decltype(DestTypeResolver(T{}));
+
+    // https://intel.github.io/llvm-docs/doxygen/classsycl_1_1__V1_1_1ext_1_1intel_1_1esimd_1_1detail_1_1simd__obj__impl.html#ad1eaca7ca43fcf87a29a9714b93d8109
+    // bit_cast_view
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h]: Fixed (removed)

File: include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h (hunk 3 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h]: Fixed (removed)

File: include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h [hunk 7]
```diff
@@ -436,6 +436,100 @@ VectorStore(uint32_t offset,
...
+    // |                             m0                           |                             m1                            | ... |                             mN                           |
+    // +----------------------------------------------------------+-----------------------------------------------------------+-----+----------------------------------------------------------+
+    //                                                                                          |
+    // __uint32_mask, simd_mask<2 * N>:                                                         V
+    // +-----------------------------+----------------------------+-----------------------------+-----------------------------+-----+-----------------------------+----------------------------+
+    // |             m0              |             m0             |             m1              |             m1              | ... |             mN              |             mN             |
+    // +-----------------------------+----------------------------+-----------------------------+-----------------------------+-----+-----------------------------+----------------------------+
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h]: Fixed.

File: test/parallel_api/experimental/radix_sort_esimd.pass.cpp (hunk 4 above)
Comment (SergeyKopienko) [reply test/parallel_api/experimental/radix_sort_esimd.pass.cpp:403]: which one?

Comment (dmitriy-sobolev) [reply test/parallel_api/experimental/radix_sort_esimd.pass.cpp:403]: Sorry, it seems we do not have such TODOs. We have `/ TODO: support double and small-size integers` which partially lost its relevance.

File: include/oneapi/dpl/experimental/kt/esimd_radix_sort_one_wg.h [hunk 8]
```diff
@@ -162,7 +162,7 @@ void one_wg_kernel(sycl::nd_item<1> idx, uint32_t n, uint32_t THREAD_PER_TG, con
         if (stage != STAGES - 1) {
             #pragma unroll
             for (uint32_t s = 0; s<PROCESS_SIZE; s+=16) {
-                slm_scatter<KeyT, 16>(
+                utils::slm_scatter<KeyT, 16>(
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/experimental/kt/esimd_radix_sort_one_wg.h]: Fixed.

File: include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h (hunk 2 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h]: This new function was removed.

**Review by dmitriy-sobolev — APPROVED** [review]
LGTM

**Pull request merged**